lib/IDE/SourceEntityWalker.cpp

```diff
@@ -633,6 +633,17 @@ ASTWalker::PreWalkAction SemaAnnotator::walkToTypeReprPre(TypeRepr *T) {
                         ReferenceMetaData(SemaReferenceKind::TypeRef, None));
       return Action::StopIf(!Continue);
     }
+  } else if (auto FT = dyn_cast<FixedTypeRepr>(T)) {
+    ValueDecl *VD = FT->getType()->getAnyGeneric();
+    if (auto DT = FT->getType()->getAs<DynamicSelfType>())
+      VD = DT->getSelfType()->getAnyGeneric();
+
+    if (VD) {
+      auto Data = ReferenceMetaData(SemaReferenceKind::TypeRef, None, /*isImplicit=*/true);
```

Comment:
It's a shame that we'll be marking `Self` as implicit here, we ought to fix that in the parser by having it produce a proper TypeRepr instead of the current hack. That being said, we could potentially work around that issue here by checking if the parent `TypeExpr` is implicit or not (the ASTWalker tracks the `Parent` node during its walk).

Reply:
hrm I guess I hadn't felt too bad about `Self` being implicit in this case:

```swift
    /*
    class=Bar=s:3foo3BarC=reference, implicit, containedBy
    constructor=init()=s:3foo3BarCACycfc=reference, call, implicit, calledBy, containedBy
    v  */
    Self()
```

just taking that value literally since `Bar` isn't explicitly written there 🤔

Reply:
I wonder if instead we should drop the implicit bit all together for both here? if anything is implicit it feels like it's the type, but the constructor isn't considered implicit in the other case:

```swift
/*
       constructor=init()=s:3foo3BarVACycfc=definition, implicit, childOf
       struct=Bar=s:3foo3BarV=definition
       v  */
struct Bar {
  /*
              staticMethod=foo()=s:3foo3BarVAaCyFZ=definition, childOf
              v  */
  static func foo() -> Self {
    /*
    struct=Bar=s:3foo3BarV=reference, containedBy
    constructor=init()=s:3foo3BarVACycfc=reference, call, calledBy, containedBy
    v  */
    Bar()
  }
  /*
              staticMethod=baz()=s:3foo3BarV3bazACyFZ=definition, childOf
              v  */
  static func baz() -> Self {
    /*
    struct=Bar=s:3foo3BarV=reference, implicit, containedBy
    constructor=init()=s:3foo3BarVACycfc=reference, call, implicit, calledBy, containedBy
    v  */
    Self()
  }
}```

Reply:
I've always taken `implicit` to mean "not spelt here". @akyrtzi / @benlangmuir in case I've had that wrong this whole time 😅

Reply:
So FWIW the current output for the `init` case is:
```
1:8 | struct/Swift | Foo | s:14swift_ide_test3FooV | Def | rel: 0
1:8 | constructor/Swift | init() | s:14swift_ide_test3FooVACycfc | Def,Impl,RelChild | rel: 1
```
ie. implicit constructor. Whether that's intended or not... I'm not sure. It could certainly just be because the expression is implicit.

Having said that, my impression has always been what I said above - ie. "not spelt here". I'm also not sure what other meaning would be useful in terms of the index data. Generally it seems useful to know whether a reference is explicitly spelt (eg. should participate in a rename) vs a reference that may be known by the compiler, but isn't actually in the source explicitly.

So my preference here would be to keep it as implicit.

Reply:
I pushed an option here for making the type implicit but the ctor call explicit, but it would be nice to have a cleaner approach if you all have any ideas

Reply:
> @bnbarham That's the index data for the implicit memberwise init decl right? That should definitely be implicit :)

Heh, indeed it was :(. The reference is not implicit:
```
5:7 | constructor/Swift | init() | s:14swift_ide_test3FooVACycfc | Ref,Call,RelCall,RelCont | rel: 1
```

We discussed this offline. We should consider `Self` similar to the typealias case. Ie. the type reference should be implicit, but the constructor should be explicit.

Reply:
sounds good, this PR currently solves that with a new bool, definitely open to better ideas for how to track that

Reply:
None from me unfortunately. Any thoughts @hamishknight?

Reply:
Yeah I think that's fine for now, I think ideally we'd try and walk the decl ref separately (or mark it explicit as @keith suggested), which I believe should avoid the need for that